crates/noirc_evaluator/src/ssa/code_gen.rs

```diff
@@ -210,9 +209,44 @@ impl IRGenerator {
         self.insert_new_struct(ident_def, values)
     }
 
-    fn codegen_identifier(&mut self, ident: &Ident) -> Value {
-        let value = self.variable_values[&ident.id].clone();
-        self.get_current_value(&value)
+    fn codegen_identifier(&mut self, ident: &Ident) -> Result<Value, RuntimeError> {
+        // Check if we have already code-gen'd the definition of this variable
+        if let Some(value) = self.variable_values.get(&ident.definition) {
+            Ok(self.get_current_value(&value.clone()))
+        } else {
+            // If we haven't, it must be a global value, like a function or builtin
+            match &ident.definition {
+                Definition::Local(id) => unreachable!(
+                    "Local variable encountered before its definition was compiled: {:?}",
+                    id
+                ),
+                Definition::Function(id) => {
+                    let id = *id;
+                    if !self.context.function_already_compiled(id) {
+                        let index = self.context.get_function_index();
+                        self.create_function(id, index)?;
+                    }
+
+                    // Link this referenced function to the current function's call graph
+                    // regardless of whether it is called later or not.
+                    let callee = self.context.get_ssafunc(id).unwrap().idx;
+                    if let Some(caller) = self.function_context {
+                        function::update_call_graph(&mut self.context.call_graph, caller, callee);
```

Comment:
If we add a function not called in the call graph, we might get some false recursive call errors
I agree that in any cases the call graph is not correct anymore because we don't necessarily know at this point what function an expression references. This is why we need this 'recursive inlining', but as I said we can add this in another PR.

Reply:
I don't think we'll ever know exactly which function a value references in the case if `if private_var { f1 } else { f2 }`, just some set of functions.
Actually, thinking about it some more, it is an error or warning to not use a variable so a function that is not called should be an error. I think we can save a more precise analysis to a later PR as well, it seems rare to me that it would make an observable difference.